vsphere/resource_vsphere_virtual_machine.go

```diff
@@ -1,2380 +1,903 @@
-							// TODO: Add ValidateFunc to confirm path exists
-							Type:     schema.TypeString,
-							Optional: true,
-						},
-
-						"bootable": &schema.Schema{
-							Type:     schema.TypeBool,
-							Optional: true,
-						},
-
-						"keep_on_remove": &schema.Schema{
-							Type:     schema.TypeBool,
-							Optional: true,
-						},
-
-						"controller_type": &schema.Schema{
-							Type:     schema.TypeString,
-							Optional: true,
-							Default:  "scsi",
-							ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
-								value := v.(string)
-								found := false
-								for _, t := range DiskControllerTypes {
-									if t == value {
-										found = true
-									}
-								}
-								if !found {
-									errors = append(errors, fmt.Errorf(
-										"Supported values for 'controller_type' are %v", strings.Join(DiskControllerTypes, ", ")))
-								}
-								return
-							},
-						},
-					},
-				},
-			},
-
-			"detach_unknown_disks_on_delete": &schema.Schema{
-				Type:     schema.TypeBool,
-				Optional: true,
-				Default:  false,
-			},
-
-			"cdrom": &schema.Schema{
-				Type:     schema.TypeList,
-				Optional: true,
-				ForceNew: true,
-				Elem: &schema.Resource{
-					Schema: map[string]*schema.Schema{
-						"datastore": &schema.Schema{
-							Type:     schema.TypeString,
-							Required: true,
-							ForceNew: true,
-						},
-
-						"path": &schema.Schema{
-							Type:     schema.TypeString,
-							Required: true,
-							ForceNew: true,
-						},
-					},
-				},
-			},
-
-			// Tagging
-			vSphereTagAttributeKey: tagsSchema(),
+		Create:        resourceVSphereVirtualMachineCreate,
+		Read:          resourceVSphereVirtualMachineRead,
+		Update:        resourceVSphereVirtualMachineUpdate,
+		Delete:        resourceVSphereVirtualMachineDelete,
+		CustomizeDiff: resourceVSphereVirtualMachineCustomizeDiff,
+		Importer: &schema.ResourceImporter{
+			State: resourceVSphereVirtualMachineImport,
 		},
+		SchemaVersion: 2,
+		MigrateState:  resourceVSphereVirtualMachineMigrateState,
+		Schema:        s,
 	}
 }
 
-func resourceVSphereVirtualMachineUpdate(d *schema.ResourceData, meta interface{}) error {
-	// flag if changes have to be applied
-	hasChanges := false
-	// flag if changes have to be done when powered off
-	rebootRequired := false
-
-	// make config spec
-	configSpec := types.VirtualMachineConfigSpec{}
-
-	if d.HasChange("vcpu") {
-		configSpec.NumCPUs = int32(d.Get("vcpu").(int))
-		hasChanges = true
-		rebootRequired = true
-	}
-
-	if d.HasChange("memory") {
-		configSpec.MemoryMB = int64(d.Get("memory").(int))
-		hasChanges = true
-		rebootRequired = true
-	}
-
-	if d.HasChange("annotation") {
-		configSpec.Annotation = d.Get("annotation").(string)
-		hasChanges = true
-	}
-
+func resourceVSphereVirtualMachineCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] %s: Beginning create", resourceVSphereVirtualMachineIDString(d))
```

Comment:
We actually send a more descriptive "Creating VM" close to around when we are actually calling `CreateVM_Task`. I think this is good - it's making the distinction that the resource is beginning its create phase, of which there are _several_ steps before the VM is actually created (it might not be created even, it might be cloned, etc).